Add macronix flash driver support.

SavinayDharmappa opened the pull request:
Patch does following things,

1. Enable internal DW EEPROM mode in SPI driver.
2.  Enable SPI driver of intel_s1000
3.Add macronix MX25UM5124G flash support

nashif commented on boards/xtensa/intel_s1000_crb/Kconfig.defconfig:
```diff
@@ -75,6 +75,24 @@ config USB_PHY_2
 	def_bool y
 endif # USB
 
+config SPI_LEGACY_API
```
that is obsolete and is not in the tree anymore

nashif commented on drivers/flash/spi_flash_mx25.h:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
2018

nashif commented on drivers/flash/spi_flash_mx25_defs.h:
```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
2018

nashif commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
2018

locomuco commented:
@SavinayDharmappa is there any difference between w25qxxd SPI driver and the macronix besides device ID?
There could be a lot of code duplication avoided with moving the exiting driver to be more generic

nashif commented:
> @SavinayDharmappa is there any difference between w25qxxd SPI driver and the macronix besides device ID?

good catch, this looks ike the same driver to me with a few minor differences that in some cases invert some of the logic leading me to believe we either introducing bugs or fixing bugs in the new driver leaving them unfixed in the existing one.

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+
+	do {
+		reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+	} while (reg & MX25_WIP_BIT);
```
same here

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+
+	do {
+		reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+	} while (reg & MX25_WEL_BIT);
```
would be nice to have a safe-guard loop counter here. It could possibly loop forever if the hw has some malfunction.

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+	struct spi_flash_data *const driver_data = dev->driver_data;
+	u8_t reg = 0;
+	int ret;
+
```
remove one extra empty line

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+	k_sem_init(&data->sem, 1, UINT_MAX);
+
+	ret = spi_flash_wb_configure(dev);
+
```
you can remove this empty line here as stated before

tbursztyka commented on drivers/flash/spi_flash_mx25.h:
```diff
@@ -0,0 +1,24 @@
+	struct spi_config spi_cfg;
+	struct k_sem sem;
+};
+
```
Since there is so few structs and all defined, you could merge the content of spi_flash_mx25_defs.h here.

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+	k_sem_take(&driver_data->sem, K_FOREVER);
+
+	reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+
```
remove that empty line

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+
+	reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+	if (!(reg & MX25_WEL_BIT)) {
+		k_sem_give(&driver_data->sem);
```
what about ret = -EIO; goto out; where out would be at line 178?

Just an idea to consider.

tbursztyka commented on drivers/flash/spi_flash_mx25.c:
```diff
@@ -0,0 +1,362 @@
+	reg = spi_flash_wb_reg_read(dev, MX25_CMD_RDSR);
+
+	if (!(reg & MX25_WEL_BIT)) {
+		k_sem_give(&driver_data->sem);
```
same as previous comment about have a out goto label

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,63 @@ config FLASH_PAGE_LAYOUT
 	help
 	  Enables API for retrieving the layout of flash memory pages.
 
+config SPI_FLASH_MX25
```
create a Kconfig.mx25 and put all this content into it. Or then the root flash Kconfig will grow and grow etc...

Of course first option becomes menuconfig and not only config.

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,63 @@ config FLASH_PAGE_LAYOUT
+	bool
+	prompt "SPI NOR Flash MACRONIX MX25"
+	depends on SPI && FLASH
+
```
add a:
if SPI_FLASH_MX25

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,63 @@ config FLASH_PAGE_LAYOUT
+config SPI_FLASH_MX25_SPI_NAME
+	string
+	prompt "SPI controller device name"
+	depends on SPI_FLASH_MX25
```
and remove all these depends on SPI_FLASH_MX25

tbursztyka commented on drivers/flash/Kconfig:
```diff
@@ -36,6 +36,63 @@ config FLASH_PAGE_LAYOUT
+	default 2097152
+	help
+	  This is the flash capacity in bytes.
+
```
add
endif # SPI_FLASH_MX25

tbursztyka commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2016 Intel Corporation.
```
2018

tbursztyka commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+	printf("==========================\n");
+
+	flash_dev = device_get_binding(CONFIG_SPI_FLASH_MX25_DRV_NAME);
+
```
remove that empty line

tbursztyka commented on tests/boards/intel_s1000_crb/src/spi_flash.c:
```diff
@@ -0,0 +1,78 @@
+	}
+
+	printk("   Attempted to write %x %x\n", buf[0], buf[1]);
+	if (flash_write(flash_dev, FLASH_TEST_REGION_OFFSET, buf,
```
add an empty line here

SavinayDharmappa committed:
board: xtensa: intel_s1000_crb: Enable DW SPI driver.

patch enables DW SPI driver for intel_s1000_crb

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
drivers: spi: Enable internal DW SPI controller EEPROM mode.

patch add internal DW SPI controller EEPROME support to the SPI
DW driver and enables on intel_s1000_crb board.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
drivers: spi: Incorporate bit filed changes in SPI Ctrl in intel_s1000

SPI DW controller used in intel_s1000 as different bitfiled for
selecting TMOD in CTRL0 than the one used in quark based micro controller.
So select the DW_SPI_CTRLR0_TMOD_SHIFT based on SOc.

SPI DW controller in intel_s1000 introduces new bit filed
called DFS(Data frame size). Which selects the data frame
length. When a command sent dfs can be selected for 8 bit serial
transfer and when address and command is sent it can be selected
for 16 bit serial transfer. It allows flash driver using SPI
driver to confiure DFS multiple times. Hence allow the driver
to configure CTRL0 multiple times

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
drivers: flash: Add macronix MX25UM5124G flash support

patch add driver to access macronix MX25UM5124G flash.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

SavinayDharmappa committed:
tests: boards: intel_s1000_crb: Add sample app to read from flash

patch add sample application to read from macronix flash. It also
configures the intel_s1000 to run at 400Mhz.

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

nashif commented:
do we still need this PR?
